mplc/multi_partner_learning.py

```diff
@@ -654,3 +655,122 @@ def init_multi_partner_learning_from_scenario(scenario, is_save_data=True):
+        super().__init__(
+            scenario.partners_list,
+            scenario.epoch_count,
+            scenario.minibatch_count,
+            scenario.dataset,
+            scenario.multi_partner_learning_approach,
+            scenario.aggregation_weighting,
+            scenario.is_early_stopping,
+            is_save_data,
+            scenario.save_folder,
+            scenario.init_model_from,
+            scenario.use_saved_weights,
+        )
+
+        self.epsilon = epsilon
+        self.K = scenario.dataset.num_classes
+        self.theta = np.array(
+            ([[np.zeros((self.K, self.K)) for _ in self.partners_list] for _ in range(self.epoch_count + 1)]))
+        self.theta_ = self.theta.copy()
+        self.theta[0] = [self.init_flip_proba() for _ in self.partners_list]
+        # self.labels_map = LabelEncoder().fit_transform([str(y) for y in scenario.dataset.y_train])
+
+    def init_flip_proba(self):
+        identity = np.identity(self.K)
+        return identity * (1 - self.epsilon) + (1 - identity) * (self.epsilon / (self.K - 1))
+
+    def fit(self):
+        # initialization of the parameters
+        partners_models = self.init_with_models()
+
+        while self.epoch_index < self.epoch_count:
+            self.split_in_minibatches()
+            self.minibatch_index = 0
+            while self.minibatch_index < self.minibatch_count:
+
+                # Evaluate and store accuracy of mini-batch start model
+                model_to_evaluate = partners_models[0]
+                model_evaluation_val_data = self.evaluate_model(model_to_evaluate, self.val_data)
+                self.score_matrix_collective_models[self.epoch_index,
+                                                    self.minibatch_index] = model_evaluation_val_data[1]
+
+                for partner_index, partner in enumerate(self.partners_list):
+                    # Reference the partner's model
+                    partner_model = partners_models[partner_index]
+                    x_batch = self.minibatched_x_train[partner_index][self.minibatch_index]
+                    y_batch = self.minibatched_y_train[partner_index][self.minibatch_index]
+
+                    self.theta_[self.epoch_index] = self.theta[self.epoch_index]  # Initialize the theta_
+                    for y in range(self.K):  # TODO ? vectorize
+                        for idx, x in enumerate(x_batch):
+                            self.theta_[self.epoch_index,
+                                        partner_index,
+                                        y,
+                                        np.argmax(y_batch[idx])] *= np.max(partner_model.predict(np.expand_dims(x,
+                                                                                                                axis=0)
+                                                                                                 )
+                                                                           )
```

Comment:
Why use `np.max`?  It should  rather be : 
`partner_model.predict(np.expand_dims(x, axis=0))[y]`

Reply:
Also why `np.exand_dims` is needed?

Reply:
np.expand_dims is need because the dimension input of a conv_layer is 4. It is for batches. so, rather than have in entry x, I use a batch of size 1, [x]